# Task 1: Some Restaurant

Name and save your Jupyter Notebook as <br>
`Task1_<your name>_<NRIC number>.ipynb`.

A restaurant offers membership for its customers. As a member, the customers are issued a unique ID numbers `IDNumber` which are integers. To help manage the identification of the customers, the manager of the restaurant opted for a low-cost solution of using a one-dimensional fixed-size array to store these ID numbers. The array is first initialized with `None` elements and the elements in this array is indexed from `0` to `(Max-1)`, where `Max` is size of the array.

To determine the index `Index` in the array for the ID numbers `IDNumber`, the manager decided to use the following formula:


$$\text{Index} \leftarrow \text{IDNumber}\mod \quad \text{Max}$$


# Task 1.1

Each line in the text file, `IDNUMBER_NAME1.CSV`, contains the ID numbers of the customers together with the customers' name.

Write program code to:

- Read ID numbers from a text file and store them in the array. For the purpose of testing the program, `Max` is to be set to the value 20. Assume different IDs will have different indices.
- Print out the contents of the array in the order in which the elements are stored in the array.

Use `IDNUMBER_NAME1.CSV` to test your program code.
<div style='text-align: right; font-family: Courier New, monospace;'>[7]</div>

In [19]:
# Task 1.1

import os
import csv

def read_file(filename):
    max_val = 20
    file_path = os.path.join("resources", filename)

    ID_array = []
    with open(file_path) as f:
        reader = csv.reader(f)
        list = [i for i in reader]
        sorted_list = sorted(list, key = lambda x: int(x[0]) % 20)
        
    print(list)
    print(sorted_list)

name = "IDNUMBER_NAME1.CSV"
read_file(name)




[['82336323', 'Alfhild�Indiana'], ['96608524', 'Pelagius�Sheila'], ['79626125', 'Klaus�Ru�a'], ['85383391', 'Gerald�Aurea'], ['76313230', 'Esyllt�Hilde'], ['99154228', 'Alix�Yeong-Su'], ['26865547', 'Zenon�Darlene'], ['56999919', 'Heard�Kazik'], ['80192615', 'Corrina�Jamal'], ['19283701', 'Puja�Fawziyya']]
[['19283701', 'Puja�Fawziyya'], ['82336323', 'Alfhild�Indiana'], ['96608524', 'Pelagius�Sheila'], ['79626125', 'Klaus�Ru�a'], ['26865547', 'Zenon�Darlene'], ['99154228', 'Alix�Yeong-Su'], ['76313230', 'Esyllt�Hilde'], ['85383391', 'Gerald�Aurea'], ['80192615', 'Corrina�Jamal'], ['56999919', 'Heard�Kazik']]


# Task 1.2

In the midst of populating the array, the manager realized that using his formula, some ID numbers could be mapped to the same index, which would lead to customers not being able to uniquely identified.

Amend your program code so that if more than one ID number is mapped to the same index, your program will  
- search sequentially from the index that has already contained an ID number for an unused location in the array,  
- get the index for this unused location the array, and  
- store the duplicate ID Number at this empty location.

Use `IDNUMBER_NAME2.csv` to test your program code.
<div style='text-align: right; font-family: Courier New, monospace;'>[4]</div>

In [ ]:
# Task 1.2

# Task 1.3

Add code to your Task 1.2 program.

The program is to:

- Take as input an ID number  
- Perform a linear search on the array and output the index of the array where the ID number was found.  

Use `IDNUMBER_NAME2.CSV` to test your program code.

Run the program three times. Use the following inputs:  

80192615, 11559181 and 79626124.
<div style='text-align: right; font-family: Courier New, monospace;'>[7]</div>  

Save your Jupyter Notebook for Task 1.

In [ ]:
# Task 1.3